Persistent Database: BPL stubs

itayx opened the pull request:
To support the persistent DB feature, BPL APIs that support the
reading/writing functionality are needed

itayx committed:
framework: bpl: add DB stubs

To support the persistent DB feature, BPL APIs that support the
reading/writing functionality are needed.
These APIs will be used to access the persistent database.

Add BPL DB APIs on both platforms that support UCI and ones that do not.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

kantera800 commented on framework/platform/bpl/include/bpl/bpl_db.h:
```diff
@@ -0,0 +1,90 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
```
```suggestion
 * SPDX-FileCopyrightText: 2020 the prplMesh contributors (see AUTHORS.md)
```

itayx replied:
Fixed

kantera800 commented on framework/platform/bpl/uci/db/bpl_db.cpp:
```diff
@@ -0,0 +1,57 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2019-2020 the prplMesh contributors (see AUTHORS.md)
```
same

itayx replied:
Fixed

kantera800 reviewed (changes requested):
you should have a bit more description in the commit message for these stubs

kantera800 reviewed (approved):
approved. 
just change to `return true` in all stubs (we don't want to fail the dummy tests).

adam1985d commented on framework/platform/bpl/include/bpl/bpl_db.h:
```diff
@@ -0,0 +1,97 @@
+ * {entry1: val1,entry2: val2,...}
+ * 
+ * @param path name of the requested section.
+ * @param values [out] JSON styled string object.
```
please remove any indication to json in the comments and implementation
1. if we eventually use JSON - the BPL is unaware of it
2. the returned value in this example is a map and not a string

itayx replied:
Updated the API in question & doxygen

adam1985d commented on framework/platform/bpl/include/bpl/bpl_db.h:
```diff
@@ -0,0 +1,97 @@
+ * @brief Get all sub-entries of a parent path.
+ * 
+ * Get all sub-entries of a given parent path in a JSON styled string object
+ * {entry1: val1,entry2: val2,...}
```
see my other comment regarding JSON

itayx replied:
Removed all mentions to JSON

adam1985d commented on framework/platform/bpl/include/bpl/bpl_db.h:
```diff
@@ -0,0 +1,97 @@
+ * @param val Initial value of the parameter.
+ * @return true on success, false otherwise.
+ */
+bool db_add_param(const std::string &path, const std::string &object, const std::string &val);
```
it's ok you didn't waste time on it now - but before this PR is merged, each API must have its own separate documentation

itayx replied:
Reworked the API in question, now it receives an unordered_map<string, string>

adam1985d commented on framework/platform/bpl/include/bpl/bpl_db.h:
```diff
@@ -0,0 +1,97 @@
+} // namespace bpl
+} // namespace beerocks
+
+#endif /* _BPL_CFG_H_ */
```
replace `_BPL_CFG_H_` with `_BPL_DB_H_`

itayx replied:
changed

adam1985d commented on framework/platform/bpl/linux/bpl_db.cpp:
```diff
@@ -0,0 +1,75 @@
+bool db_add_param(const std::string &path, const std::string &object, const std::string &val)
+{
+    LOG(TRACE) << __func__ << " - NOT IMPLEMENTED!";
+    return false;
```
# for all stubs #
please return true,
I know this hides the point that this function is called but isn't implemented but we are going to add the implementation now so "true" should be returned.

itayx replied:
Updated

adam1985d commented on framework/platform/bpl/uci/db/bpl_db.cpp:
```diff
@@ -0,0 +1,75 @@
+    return false;
+}
+
+bool db_set_param(const std::string &path, const std::string &object, const int &val)
```
for simple data types `set` operation no need for `const ref&` - a copy is ok

itayx replied:
Changed to use an unordered_map<string, string>, therefore using a reference is better since we're working with a container and do not want to copy the whole thing

morantr commented on framework/platform/bpl/include/bpl/bpl_db.h:
```diff
@@ -0,0 +1,97 @@
+ * 
+ * @param path name of the requested section.
+ * @param object full name of the parameter.
+ * @param val [in] value of the given parameter.
```
For const reference need to write the oxygen like this;
`@param[in] param_name` and not `@param param_name [in]`

Also, it is required to add the direction for all  **non-const** references, so here it is not required.
Anyway, if you decide to do it also to **const** reference arguments, it is probably best if you do it on all of them in the file.

This comment apply to all the changes you've done in this commit.

itayx replied:
updated the oxygen according to the rest of the BPL apis

morantr commented on framework/platform/bpl/include/bpl/bpl_db.h:
```diff
@@ -0,0 +1,97 @@
+/****************************************************************************/
+
+/**
+ * @brief Add parameter to the presistent DB
```
Missing dot at sentence end.

itayx replied:
Done

adam1985d reviewed (changes requested):
the PR is mostly OK but please address each of the comments I provided and let me re-review

morantr commented on framework/platform/bpl/include/bpl/bpl_db.h:
```diff
@@ -0,0 +1,97 @@
+ * @return true on success, false otherwise.
+ */
+bool db_add_param(const std::string &path, const std::string &object, const std::string &val);
+bool db_add_param(const std::string &path, const std::string &object, const int &val);
```
There is no point passing `int` to a function as a const reference.  We pass parameters as const reference only if it's size is or might be greater than sizeof(int), which in our case is 4 bytes.
Apply to other functions in the file as well.

itayx replied:
This is no longer relevant, but I understood the point, thanks.

mariomaz commented on framework/platform/bpl/include/bpl/bpl_db.h:
```diff
@@ -0,0 +1,87 @@
+ * 
+ * @param[in] entry_type type of the requested entry.
+ * @param[in] entry_name name of the requested entry.
+ * @param[in] values initial values of the entry.
```
Please document what are the keys and values for the map.
If keys are the entry name then the function should accept a vector and not a map

itayx replied:
see my comment https://github.com/prplfoundation/prplMesh/pull/1460#issuecomment-651028711

mariomaz commented on framework/platform/bpl/include/bpl/bpl_db.h:
```diff
@@ -0,0 +1,87 @@
+ * 
+ * @param[in] entry_type type of the requested entry.
+ * @param[in] entry_name name of the requested entry.
+ * @param[in] values unordered_map containing the values the requested entry, to be set.
```
values should be a vector, not a map.

itayx replied:
see my comment https://github.com/prplfoundation/prplMesh/pull/1460#issuecomment-651028711

mariomaz commented on framework/platform/bpl/include/bpl/bpl_db.h:
```diff
@@ -0,0 +1,87 @@
+ * @param[in] values unordered_map containing the values the requested entry, to be set.
+ * @return true on success, false otherwise.
+ */
+bool db_set_entry_params(const std::string &entry_type, const std::string &entry_name,
```
The names of the API functions shall be consistent. I mean if `db_add_entry` is used, then this function shall be `db_set_entry`

itayx replied:
see my comment https://github.com/prplfoundation/prplMesh/pull/1460#issuecomment-651028711

mariomaz commented on framework/platform/bpl/include/bpl/bpl_db.h:
```diff
@@ -0,0 +1,87 @@
+ * @param [in/out] values unordered_map containing the values of the requested entry.
+ * @return true on success, false otherwise.
+ */
+bool db_get_entry_params(const std::string &entry_type, const std::string &entry_name,
```
values should be a vector, not a map.

itayx replied:
see my comment https://github.com/prplfoundation/prplMesh/pull/1460#issuecomment-651028711

mariomaz commented on framework/platform/bpl/include/bpl/bpl_db.h:
```diff
@@ -0,0 +1,87 @@
+ * @param [in/out] values unordered_map containing the values of the requested entry.
+ * @return true on success, false otherwise.
+ */
+bool db_get_entry_params(const std::string &entry_type, const std::string &entry_name,
```
The names of the API functions shall be consistent. I mean if `db_add_entry` is used, then this function shall be `db_get_entry`

itayx replied:
see my comment https://github.com/prplfoundation/prplMesh/pull/1460#issuecomment-651028711

mariomaz commented on framework/platform/bpl/include/bpl/bpl_db.h:
```diff
@@ -0,0 +1,87 @@
+ * @param[in] values unordered_map containing the values the requested entry, to be set.
+ * @return true on success, false otherwise.
+ */
+bool db_set_entry_params(const std::string &entry_type, const std::string &entry_name,
```
What is the `entry_type` used for? If the `entry_name` is unique then no other value is required for disambiguation to look up the entry.
This comment applies to the rest of functions below too.

itayx replied:
see my comment https://github.com/prplfoundation/prplMesh/pull/1460#issuecomment-651028711

mariomaz commented on framework/platform/bpl/linux/bpl_db.cpp:
```diff
@@ -0,0 +1,56 @@
+bool db_add_entry(const std::string &entry_type, const std::string &entry_name,
+                  const std::unordered_map<std::string, std::string> &values)
+{
+    LOG(TRACE) << __func__ << " - NOT IMPLEMENTED!";
```
I think `__func__` is not needed because `LOG` is already writing the function name

itayx replied:
removed

mariomaz reviewed (approved):
Choosing sensible names for API functions can drastically smooth out the learning curve for new developers, helping them intuitively know what to look for and where to find it.

Consistency is key when naming functions and parameters. No matter how well implemented the API is, your API will always feel clumsy if names are inconsistent. .

Last but not least, choose the most appropriate parameter types in each occasion. 

My comments are very easy to fix in case you want to do it, so approving.

itayx commented:
@mariomaz I have a few notes regarding your comments:
- The `entry_name` is only partly unique and the only restriction is that entries of the same type are forced to use unique name.
- Each entry will contain parameters within in a pair `<name,value>`.
And each entry will probably contain multiple parameters which are differentiated by their parameter name
Therefore to use the DB an unordered_map is needed, though I understand your point to the argument naming
So I'm changing the "values" argument to "param" so it will be more sensible

itayx committed:
cppcheck: bpl issues update

Currently the APIs are not yet implemented which causes the cppcheck to
mark them as issues, but once they are implemented the issues will not
appear.

Add known BPL issues, that will be removed once the BPL DB functions
will be implemented.

Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>
